### PR TITLE
GTFS profile & routing 

### PR DESCRIPTION
**Pull request by takb**
### Pull Request Checklist

- [x] 1. I have [**rebased**][rebase] the latest version of the master branch into my feature branch and all conflicts
         have been resolved.
- [x] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the
         [Unreleased] heading.
- [ ] 3. I have documented my code using JDocs tags.
- [x] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [ ] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [x] 6. I have created API tests for any new functionality exposed to the API.
- [] 7. If changes/additions are made to the ors-config.json file, I have added these to the [ors config documentation][config]
         along with a short description of what it is for, and documented this in the Pull Request (below).
- [x] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [ ] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [ ] 10. For new features or changes involving building of graphs, I have tested on a larger dataset
          (at least Germany), and the graphs build without problems (i.e. no out-of-memory errors).
- [ ] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the
          importer etc.), I have generated longer distance routes for the affected profiles with different options
          (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end
          points generated from the current live ORS.
          If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [x] 12. I have written in the Pull Request information about the changes made including their intended usage
          and why the change was needed.
- [x] 13. For changes touching the API documentation, I have tested that the API playground [renders correctly][api].

### Information about the changes
- Key functionality added:
GTFS routing support. 

Documentation needs to be updated before we roll out the next version. @koebi could maybe also help me check if the branch feature-gtfs  is fully superseded by this, please.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -34,6 +34,8 @@ RELEASING:
  -->
 
 ## [Unreleased]
+### Added
+- GTFS support
```
Comment (MichaelsJP) [inline CHANGELOG.md]:
```suggestion
- GTFS support for routing  ([#1316](https://github.com/GIScience/openrouteservice/pull/1316))
```

File: .github/workflows/run_maven_tests.yml [hunk 2]
```diff
@@ -35,7 +35,7 @@ jobs:
     - name: Copy ors test config for API tests
       run: cp ./openrouteservice-api-tests/conf/ors-config-test.json ./openrouteservice/src/main/resources/ors-config.json
     - name: Test, build, run API tests
-      run: mvn -B -f openrouteservice/pom.xml verify jacoco:report
+      run: export MAVEN_OPTS="-Xmx3000m -Xms2500m" && mvn -B -f openrouteservice/pom.xml verify jacoco:report
```
Comment (MichaelsJP) [inline .github/workflows/run_maven_tests.yml]: We should probably use a smaller test set?

File: openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/routing/ResultTest.java [hunk 3]
```diff
@@ -3651,6 +3653,177 @@ public void expectMaxpeedHgvForward() {
...
+            .body("features[0].properties.legs[0].containsKey('instructions')", is(true))
+            .body("features[0].properties.legs[0].containsKey('geometry')", is(true))
+            .body("features[0].properties.legs[1].containsKey('feed_id')", is(true))
+            .body("features[0].properties.legs[1].containsKey('trip_id')", is(true))
+            .body("features[0].properties.legs[1].containsKey('route_id')", is(true))
+            .body("features[0].properties.legs[1].containsKey('route_type')", is(true))
+            .body("features[0].properties.legs[1].containsKey('route_desc')", is(true))
+            .body("features[0].properties.legs[1].containsKey('geometry')", is(true))
+            .statusCode(200).extract().response();
+    }
```
Comment (MichaelsJP) [inline openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/routing/ResultTest.java]:
```suggestion
    @Test
    public void testPTJSON() {
        JSONArray coordinates =  new JSONArray();
        JSONArray coord1 = new JSONArray();
        coord1.put(8.6729581);
        coord1.put(49.4468535);
        coordinates.put(coord1);
        JSONArray coord2 = new JSONArray();
        coord2.put(8.7067204);
        coord2.put(49.3786147);
        coordinates.put(coord2);
        JSONObject body = new JSONObject();
        body.put("coordinates", coordinates);
        body.put("instructions", true);
        body.put("elevation", true);
        body.put("departure", "2022-07-04T13:02:26Z");
        body.put("walking_time", "PT30M");
        Response res = given()
            .header("Accept", "application/json")
            .header("Content-Type", "application/json")
            .pathParam("profile", getParameter("ptProfile"))
            .body(body.toString())
            .when()
            .post(getEndPointPath() + "/{profile}")
            .then().log().ifValidationFails()
            .assertThat()
            .body("any { it.key == 'routes' }", is(true))
            .body("routes.size()", is(3))
            .body("routes[0].summary.transfers", is(2))
            .body("routes[0].legs.size()", is(5))
            .body("routes[0].legs[0].containsKey('arrival')", is(true))
            .body("routes[0].legs[0].containsKey('departure')", is(true))
            .body("routes[0].legs[0].containsKey('instructions')", is(true))
            .body("routes[0].legs[0].containsKey('geometry')", is(true))
            .body("routes[0].legs[1].containsKey('feed_id')", is(true))
            .body("routes[0].legs[1].containsKey('trip_id')", is(true))
            .body("routes[0].legs[1].containsKey('route_id')", is(true))
            .body("routes[0].legs[1].containsKey('route_type')", is(true))
            .body("routes[0].legs[1].containsKey('route_desc')", is(true))
            .body("routes[0].legs[1].containsKey('geometry')", is(true))
            .statusCode(200).extract().response();
    }

    @Test
    public void testPTGeoJSON() {
        JSONArray coordinates =  new JSONArray();
        JSONArray coord1 = new JSONArray();
        coord1.put(8.6729581);
        coord1.put(49.4468535);
        coordinates.put(coord1);
        JSONArray coord2 = new JSONArray();
        coord2.put(8.7067204);
        coord2.put(49.3786147);
        coordinates.put(coord2);
        JSONObject body = new JSONObject();
        body.put("coordinates", coordinates);
        body.put("instructions", true);
        body.put("elevation", true);
        body.put("departure", "2022-07-04T13:02:26Z");
        body.put("walking_time", "PT30M");
        Response res = given()
            .header("Accept", "application/geo+json")
            .header("Content-Type", "application/json")
            .pathParam("profile", getParameter("ptProfile"))
            .body(body.toString())
            .when()
            .post(getEndPointPath() + "/{profile}/geojson")
            .then().log().ifValidationFails()
            .assertThat()
            .body("any { it.key == 'features' }", is(true))
            .body("features.size()", is(3))
            .body("features[0].properties.transfers", is(2))
            .body("features[0].properties.legs.size()", is(5))
            .body("features[0].properties.legs[0].containsKey('arrival')", is(true))
            .body("features[0].properties.legs[0].containsKey('departure')", is(true))
            .body("features[0].properties.legs[0].containsKey('instructions')", is(true))
            .body("features[0].properties.legs[0].containsKey('geometry')", is(true))
            .body("features[0].properties.legs[1].containsKey('feed_id')", is(true))
            .body("features[0].properties.legs[1].containsKey('trip_id')", is(true))
            .body("features[0].properties.legs[1].containsKey('route_id')", is(true))
            .body("features[0].properties.legs[1].containsKey('route_type')", is(true))
            .body("features[0].properties.legs[1].containsKey('route_desc')", is(true))
            .body("features[0].properties.legs[1].containsKey('geometry')", is(true))
            .statusCode(200).extract().response();
    }
```

File: openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/routing/ResultTest.java [hunk 4]
```diff
@@ -3651,6 +3653,177 @@ public void expectMaxpeedHgvForward() {
...
+
+    @Test
+    public void testPTFail() {
```
Comment (MichaelsJP) [inline openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/routing/ResultTest.java]: Either split in multiple tests or parametrize the tests.

**Review by MichaelsJP — CHANGES_REQUESTED** [review]

File: .github/workflows/run_maven_tests.yml (hunk 2 above)
Comment (takb) [reply .github/workflows/run_maven_tests.yml]: done

File: openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/routing/ResultTest.java (hunk 4 above)
Comment (takb) [reply openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/routing/ResultTest.java]: Parametrized.

**Comment by TheGreatRefrigerator** [comment]
some issues with the gtfs documentation:
- [x] `pt` profile should prob. be `public-transport` as it doesn't fit with the rest.
- [ ] `pt` profile should only be used with directions (if not implemented for isos and matrix). Not sure if we can restrict profiles for specific endpoints yet
- [x] default and example values for `ignore_transfers`, `schedule`, `schedule_duration`, `schedule_rows` and `walking_time` should be meaningful. Default values should be preferred over example values so the info "what is set when the param is not passed" is there.
- [x] the description(`value` in `ApiModelProperty`) for the above parameters doesn't really make clear what happens if you set them. Probably a bit more explanation on what they do.
- [x] the description sometimes contains `public transit` code always names things `public_transport` (except for `PUBLIC_TRANSIT` in APIEnums.java) -> use `public_transport` consistently and get rid of public transit, also in comments

**Comment by koebi** [comment]
- [x] add config documentation for gtfs profile
- [ ] add info that transfers will be calculated by a set foot profile

File: openrouteservice/src/main/java/org/heigit/ors/api/responses/routing/json/JSONLeg.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+import org.locationtech.jts.geom.Coordinate;
+
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.List;
+
+@ApiModel(value="JSONLeg", description = "Leg of a route")
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public class JSONLeg {
+    @ApiModelProperty(value = "The type of the leg, possible values are currently 'walk' and 'pt'.", example = "pt")
```
Comment (MichaelsJP) [inline openrouteservice/src/main/java/org/heigit/ors/api/responses/routing/json/JSONLeg.java:38]: Shouldn't we better rename all `pt` instances to `public-transit` ?

File: openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/routing/ResultTest.java [hunk 6]
```diff
@@ -110,6 +140,7 @@ public ResultTest() {
         addParameter("bikeProfile", "cycling-regular");
         addParameter("carProfile", "driving-car");
         addParameter("footProfile", "foot-walking");
+        addParameter("ptProfile", "pt");
```
Comment (MichaelsJP) [inline openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/routing/ResultTest.java]:
```suggestion
        addParameter("ptProfile", "public-transit");
```

File: openrouteservice/src/main/java/org/heigit/ors/routing/RoutingProfileType.java [hunk 7]
```diff
@@ -192,6 +194,9 @@ public static int getFromString(String profileType) {
             case "wheelchair":
                 return WHEELCHAIR;
 
+            case "pt":
```
Comment (MichaelsJP) [inline openrouteservice/src/main/java/org/heigit/ors/routing/RoutingProfileType.java]:
```suggestion
            case "public-transit":
```

**Review by MichaelsJP — CHANGES_REQUESTED** [review]

File: openrouteservice/src/main/java/org/heigit/ors/api/responses/routing/json/JSONLeg.java (hunk 5 above)
Comment (takb) [reply openrouteservice/src/main/java/org/heigit/ors/api/responses/routing/json/JSONLeg.java:38]: image
@koebi renamed `pt`  to `public-transport` and it seems okay to me...

**Comment by takb** [comment]
@koebi says he's done with docs and checks, I think this is ready @MichaelsJP

**Review by MichaelsJP — APPROVED** [review]

**Pull request merged**